src/software/ai/navigator/path_planner/hrvo/hrvo_agent.h

```diff
@@ -0,0 +1,217 @@
+                         Duration time_step) override;
+
+
+    /**
+     * Computes the new velocity of this agent.
+     *
+     * @param robots a map from robot ids to robots, for all robots in the simulation
+     */
+    void computeNewVelocity(const std::map<unsigned int, std::shared_ptr<Agent>> &robots,
+                            Duration time_step) override;
+
+    /**
+     * Computes the preferred velocity of this agent.
+     *
+     * @param time_step
+     * @return the computed preferred velocity
+     */
+    Vector computePreferredVelocity(Duration time_step) override;
+
+    /**
+     * Compute all the velocity obstacles that this Agent should take into account and
+     * add it to `velocityObstacles`.
+     *
+     * @param the robots nearest neighbours
+     */
+    void computeVelocityObstacles(
+        const std::map<RobotId, std::shared_ptr<Agent>> &robots);
+
+    /**
+     * Create the VO for the given agent, relative to this agent
+     *
+     * @param other_agent The Agent which this velocity obstacle is being generated for
+     * @return The velocity obstacle which other_agent should see for this Agent
+     */
+    VelocityObstacle createVelocityObstacle(const Agent &other_agent) override;
+    /**
+     * compute the neighbours of this robot
+     * @param robots
+     * @return the robot simulator ids
+     */
+    std::vector<RobotId> computeNeighbors(
+        const std::map<RobotId, std::shared_ptr<Agent>> &robots);
+
+    /**
+     * get the velocity obstacles that this agent sees
+     * @return a list of velocity obstacles
+     */
+    std::vector<VelocityObstacle> getVelocityObstacles();
+
+    /**
+     * get the velocity obstacle for the ball that the robot sees
+     * @return an o obstacle pointer
+     */
+    std::optional<ObstaclePtr> getBallObstacle();
+
+
+    /**
+     * construct the visualization proto for this robot
+     *
+     * @param friendly_team_colour the team color for this robot
+     */
+    void visualize(TeamColour friendly_team_colour);
+
+
+    // This class holds the calculated candidate velocity given two velocity obstacles
+    class CandidateVelocity
+    {
+       public:
+        /**
+         * Constructor. all candidate velocities occur at the intersections of vo's.
+         * @param velocity the velocity for this candidate
+         * @param index_1 the index of the first intersecting velocity obstacle
+         * @param index_2 the index of the second intersecting velocity obstacle
+         */
+        explicit CandidateVelocity(Vector velocity, int index_1, int index_2)
+            : velocity(velocity), obstacle_indexes(std::pair(index_1, index_2))
+        {
+        }
+
+        // The velocity of the candidate.
+        Vector velocity;
+        // pair of (potentially) intersecting velocity obstacles
+        // from which we can calculate the candidate velocity
+        std::pair<int, int> obstacle_indexes;
+    };
```

Comment:
in the future, not blocking, but we could use a shared_ptr to the actual ObstaclePtrs that intersect